akka-projection-grpc/src/main/scala/akka/persistence/EventWriter.scala

```diff
@@ -0,0 +1,339 @@
+                  case Some(stateForPid) =>
+                    stateForPid.waitingForReply.get(sequenceNr) match {
+                      case None =>
+                        throw new IllegalStateException(
+                          s"Got write success reply for event with no waiting request, probably a bug (pid $pid, seq nr $sequenceNr)")
+                      case Some((_, waiting)) =>
+                        if (context.log.isTraceEnabled)
+                          context.log.trace2(
+                            "Successfully wrote event persistence id [{}], sequence nr [{}]",
+                            pid,
+                            message.sequenceNr)
+                        waiting ! StatusReply.success(WriteAck(pid, sequenceNr))
+                        val newState = stateForPid.copy(waitingForReply = stateForPid.waitingForReply - sequenceNr)
+                        handleUpdatedStateForPid(pid, newState)
+                        Behaviors.same
+                    }
+                }
+
+              case JournalProtocol.WriteMessageFailure(message, error, `actorInstanceId`) =>
+                val pid = message.persistenceId
+                val sequenceNr = message.sequenceNr
+                perPidWriteState.get(pid) match {
+                  case None =>
+                    throw new IllegalStateException(
+                      s"Got error reply for event with no waiting request, probably a bug (pid $pid, seq nr $sequenceNr)",
+                      error)
+                  case Some(state) =>
+                    // write failure could be re-delivery, we need to check
+                    state.waitingForReply.get(sequenceNr) match {
+                      case None =>
+                        throw new IllegalStateException(
+                          s"Got error reply for event with no waiting request, probably a bug (pid $pid, seq nr $sequenceNr)",
+                          error)
+                      case Some(_) =>
+                        // quite likely a re-delivery of already persisted events, check highest seqnr
+                        val sortedSeqNrs = state.waitingForReply.keys.toSeq.sorted
+                        val maxSeqNrFinderName = URLEncoder.encode(s"MaxSeqNrFinder-$pid", "UTF-8")
+                        if (context.child(maxSeqNrFinderName).isEmpty) {
+                          // first failure in batch, but batch is atomic so we know it all failed
+                          context.spawn[Nothing](
+                            MaxSeqNrFinder(
+                              context.self,
+                              journal,
+                              pid,
+                              sortedSeqNrs.head,
+                              sortedSeqNrs.last,
+                              error.getMessage),
+                            maxSeqNrFinderName)
+                        }
+                        Behaviors.same
```

Comment:
We trigger the check for the first `WriteMessageFailure` but for a batch of events there will be one WriteMessageFailure for each event (same pid). How do we handle that? We only want to do one lookup, and also for the partial case we send a new Write to the journal before we are sure that we have handled all replies from the previous (for that pid).

Reply:
I think it's time to write a test for the EventWriter. Maybe two, one with a real journal (in-mem testkit), and one with a fake responding to the journal protocol from the test to easier simulate different error scenarios.

Reply:
Yes, definitely need to add tests for the event writer in isolation.

I'm using the presence of the child as a flag that we already started checking the batch to ignore subsequent failure events for that pid until that has completed (and a new write is triggered).

Reply:
I missed that part. It could take a while until this actor sees that the child has been terminated and removed the child name is removed here so maybe that could be a theoretical problem that a fail reply for next batch arrives before the child has been removed.

Reply:
Better with a flag in the per-pid-state then.

Reply:
Changed in [9799279](https://github.com/akka/akka-projection/pull/932/commits/9799279daa99c9b3fae1d3ab2219f106d221fc0f)